src/utils.ml

```diff
@@ -203,13 +203,26 @@ module Cached_digest = struct
     ; table       = Hashtbl.create 1024
     }
 
+  let dir_digest stat =
+    let ints = stat.Unix.st_size + stat.st_perm in
+    (string_of_int ints ^
+     string_of_float stat.st_mtime ^
+     string_of_float stat.st_ctime)
```

Comment:
It's a relatively minor thing, but (a ^ b ^ c) will allocate an intermediate string (a^b) that creates unnecessary garbage. It might be better to do a sprintf here to do it in one allocation if the digest function is called a lot

Reply:
Now that you point that out, I just realise that we are missing separators in this string. i.e. two different stats records might produce the same string... That's very unlikely, but just to avoid readers wondering about it, we should add separators

Reply:
Fix here https://github.com/ocaml/dune/pull/1592

Reply:
thank you all!